### ISSUE TITLE
Python port

### ISSUE THREAD

**Issue by srush** [bijection/g9#6]
Hi there, 

I really love the g9.js demos! I think they convey how gradients work in such a nice and intuitive manner. 

Just for fun I was playing with the idea of running a Torch-like library in the browser. I ported over most of the g9 code to python and ran it with AD (as opposed to finite-differences) -> https://srush.github.io/g9py/ . Not sure why this is better, but it might let you run models with tons of parameters. Let me know if this is okay.

Cheers!
Sasha

**Comment by bijection** [comment]
Very cool! I'd be happy to add a link to this if you like.

It seems at first click that it's a bit less performant for the recursive demos - I haven't read through the code yet, but did you implement the pruning part that uses `ctx.pure`? Or does that not play nice with AD?

**Comment by srush** [comment]
Right, yeah I forget why I didn't add that. Need to think about what it means for with torch-style AD.
